src/decode.rs

```diff
@@ -769,9 +769,14 @@ unsafe fn read_pal_plane(
     let used_cache = &used_cache[..i];
 
     // parse new entries
+    #[allow(unused_mut)]
+    let mut pal_guard;
     let pal = if t.frame_thread.pass != 0 {
-        &mut f.frame_thread.pal[(((t.by >> 1) + (t.bx & 1)) as isize * (f.b4_stride >> 1)
-            + ((t.bx >> 1) + (t.by & 1)) as isize) as usize][pli]
+        pal_guard = f.frame_thread.pal.index_mut(
+            (((t.by >> 1) + (t.bx & 1)) as isize * (f.b4_stride >> 1)
+                + ((t.bx >> 1) + (t.by & 1)) as isize) as usize,
+        );
+        &mut pal_guard[pli]
     } else {
         &mut t.scratch.c2rust_unnamed_0.pal[pli]
     };
```

Comment:
We don't need a non-transient guard here.  We can do the 128-bit write at the end instead of holding the `&mut` the whole time:

```rust
    let mut pal = [0; 8];
    if i < pal_sz {
        let mut prev = rav1d_msac_decode_bools(&mut ts.msac, f.cur.p.bpc as u32) as u16;
        pal[i] = prev;
        i += 1;

        if i < pal_sz {
            let mut bits = f.cur.p.bpc as u32 + rav1d_msac_decode_bools(&mut ts.msac, 2) - 3;
            let max = (1 << f.cur.p.bpc) - 1;

            loop {
                let delta = rav1d_msac_decode_bools(&mut ts.msac, bits) as u16;
                prev = cmp::min(prev + delta + not_pl, max);
                pal[i] = prev;
                i += 1;
                if prev + not_pl >= max {
                    pal[i..].fill(max);
                    break;
                } else {
                    bits = cmp::min(bits, 1 + ulog2((max - prev - not_pl) as u32) as u32);
                    if !(i < pal_sz) {
                        break;
                    }
                }
            }
        }

        // merge cache+new entries
        let mut n = 0;
        let mut m = used_cache.len();
        for i in 0..pal_sz {
            if n < used_cache.len() && (m >= pal_sz || used_cache[n] <= pal[m]) {
                pal[i] = used_cache[n];
                n += 1;
            } else {
                pal[i] = pal[m];
                m += 1;
            }
        }
    } else {
        pal[..used_cache.len()].copy_from_slice(&used_cache);
    }

    if t.frame_thread.pass != 0 {
        f.frame_thread.pal.index_mut(
            (((t.by >> 1) + (t.bx & 1)) as isize * (f.b4_stride >> 1)
                + ((t.bx >> 1) + (t.by & 1)) as isize) as usize,
        )[pli] = pal;
    } else {
        t.scratch.c2rust_unnamed_0.pal[pli] = pal;
    };
```

Reply:
It's not obviously clear to me that this is the same without copying the data out of the array first. It might be, but I'd have to think on this. What's wrong with the mutable reference here for now?

Reply:
To me, it's simpler code.  We create `pal` and then write it, so it's easier to reason about things.  We don't have to reason about lifetimes.  `pal` is only written to, so this should be fine to do.

Reply:
It's not quite write-only, `pal[m]` is read from and I'm having trouble convincing myself that every entry is written to (i.e. the original contents are never read). I'd prefer a simpler diff here, considering that the "guard" is zero-cost in release mode.

Reply:
> It's not quite write-only, `pal[m]` is read from and I'm having trouble convincing myself that every entry is written to (i.e. the original contents are never read). I'd prefer a simpler diff here, considering that the "guard" is zero-cost in release mode.

Hmm, yeah, that's not as trivial I thought.  Nevermind then.